feat: Add event registration

bot190 opened the pull request:
Adds the ability for users to register for events. Registered users are shown on the event stats page visible to mentors. 

This mechanism is intended to support multiple time blocks for each event. Currently a single block for the whole duration is automatically created. 

Each block tracks whether the user is registered, and a comment. The comment field is intended for indicating if someone is arriving late, leaving early, etc.

Users names will show up as links on an event page if they have left a comment for the block. Clicking the link opens a modal with their comment.

Work Remaining:

* ~~Re-work event stats page to only list users actively registered, not all students that have event registrations created.~~
* ~~Ensure comments are propagated to the form if a user goes back to the registration link.~~
* Make registration pages more visible.

bot190 committed:
feat: Add event registration

Adds the ability for users to register for events. Registered users are shown on the event stats page visible to mentors. 

This mechanism is intended to support multiple time blocks for each event. Currently a single block for the whole duration is automatically created. 

Each block tracks whether the user is registered, and a comment. The comment field is intended for indicating if someone is arriving late, leaving early, etc.

Users names will show up as links on an event page if they have left a comment for the block. Clicking the link opens a modal with their comment.

Adds a new page visible to all users showing events that are open for registration and a link to the registration page for each event.

msoucy commented on signinapp/__init__.py:
```diff
@@ -222,30 +222,35 @@ def init_default_db():
             start=now,
             end=now + offset,
-            type_=db.session.scalar(select(EventType).filter_by(name="Training")),
+            event_type=db.session.scalar(select(EventType).filter_by(name="Training")),
```
Can we get an `EventType.from_name` or similar for this?

msoucy commented on signinapp/events.py:
```diff
@@ -214,12 +235,18 @@ def bulk_events():
+            location=form.location.data,
+            start=form.start.data,
+            end=form.end.data,
+            type_=event_type,
```
Isn't this supposed to be called `event_type` in the `create_event` function?

msoucy commented on signinapp/events.py:
```diff
@@ -266,3 +293,48 @@ def search():
+
+@events.route("/events/register", methods=["GET", "POST"])
+def register_event():
+    event: List[Event] = db.session.get(Event, request.args["event_id"])
```
You don't need `from typing import List`, you can use `list` directly as of [Python 3.9](https://docs.python.org/3/whatsnew/3.9.html)

msoucy commented on signinapp/templates/event_register.html.jinja2:
```diff
@@ -0,0 +1,31 @@
+{% extends "base.html.jinja2" %}
```
No way to get to this page?

bot190 replied:
Yeah, I was thinking of adding a new event page that students can see enabled events at, though I need to see how we're using that field. Or we could just rely upon a mentor sharing the link which is likely how students will see it anyway.

msoucy replied:
I'd personally prefer to give the students a "dashboard"/way to see all upcoming events that are able to be registered for

msoucy commented on signinapp/templates/event_register.html.jinja2:
```diff
@@ -0,0 +1,31 @@
+              <th scope="col">Comment</th>
+            </tr>
+          </thead>
+          <tbody id=eventblocksbody">
```
Missing quote here

msoucy commented on signinapp/model.py:
```diff
@@ -497,6 +527,10 @@ class EventType(db.Model):
     events: list[Event] = db.relationship("Event", back_populates="type_")
 
+    @staticmethod
+    def from_name(name: str):
```
Missing return type

msoucy commented on signinapp/model.py:
```diff
@@ -474,6 +475,35 @@ def scan(self, user_code) -> StampEvent:
+
+        # Add default block for the entire event time
+        block = EventBlock(start=start, end=end, event_id=ev.id)
+        db.session.add(block)
```
Do we not need to flush again after this?

bot190 replied:
The caller does a commit, no need to separately flush here. We only need to for the event so that it gets an ID.

msoucy commented on signinapp/events.py:
```diff
@@ -165,17 +188,35 @@ def event_stats():
+        blocks[block] = sorted(blocks[block], key=lambda registration: registration[0])
+
+    blocks = sorted(
+        ((block, names) for block, names in blocks.items()),
```
Is this generator actually needed? It looks like it could just be replaced with `blocks.items()`

msoucy commented on signinapp/model.py:
```diff
@@ -474,6 +475,35 @@ def scan(self, user_code) -> StampEvent:
         return StampEvent(user.human_readable, "in")
 
+    @staticmethod
+    def create_event(
```
Smurf function name

msoucy commented on signinapp/model.py:
```diff
@@ -653,3 +687,60 @@ class Subteam(db.Model):
+
+    # Link to event block
+    event_block_id = db.Column(db.Integer, db.ForeignKey("eventblocks.id"))
+    event_block = db.relationship("EventBlock", back_populates="registrations")
```
Please add type annotations to the relationships

msoucy commented on signinapp/templates/event_stats.html.jinja2:
```diff
@@ -49,4 +82,33 @@
+        <button type="button" class="btn-close btn-close-white" data-bs-dismiss="modal" aria-label="Close"></button>
+      </div>
+      <div class="modal-body" >
+        TEST
```
?

msoucy commented on signinapp/__init__.py:
```diff
@@ -215,30 +214,35 @@ def init_default_db():
             start=now,
             end=now + offset,
-            type_=db.session.scalar(select(EventType).filter_by(name="Training")),
+            event_type=EventType.from_name(name="Training"),
```
Since you're making a factory method anyways, it would be kinda nice to have `event_type` take `str | EventType`, and call `Event.from_name` if it's a string. Low priority though.

msoucy committed:
fix: Set registration open flag in Event create function

msoucy commented on signinapp/events.py:
```diff
@@ -159,23 +186,57 @@ def list_upcoming_events():
 
+@events.route("/events/open")
+def list_open_events():
+    events: list[Event] = db.session.execute(
```
I'm curious if there's a difference between `db.session.execute(...).scalars()` and `db.session.scalars(...)`

bot190 replied:
They're equivalent, but `db.session.scalars(...)` is obviously shorter. I've updated all the calls in events.py (didn't see any others) to use this form.

msoucy commented on signinapp/events.py:
```diff
@@ -159,23 +186,57 @@ def list_upcoming_events():
+    events: list[Event] = db.session.execute(
+        select(Event)
+        .filter_by(enabled=True, registration_open=True)
+        .order_by(Event.start)
```
Do we expect to have overlapping time blocks at all? If so, this should `.order_by(Event.start, Event.end)`.

Also if you do `.order_by(Event.start.desc())` then you don't need to reverse the list inside the template

msoucy commented on signinapp/events.py:
```diff
@@ -268,5 +333,53 @@ def search():
+                registered=block["register"],
+            )
+        db.session.commit()
+        flash(f"Successfully Signed up for {event.name}")
```
Should `Signed` be capitalized?

msoucy commented on signinapp/templates/event_stats.html.jinja2:
```diff
@@ -2,6 +2,12 @@
+<script>
+  function copyRegistrationLink() {
+    navigator.clipboard.writeText("{{registration_url}}");
+    alert("Copied: {{registration_url}}");
```
Let's not use `alert`... I'd be fine with using another toast method like we do on the scan page though.

msoucy commented on signinapp/templates/event_stats.html.jinja2:
```diff
@@ -12,6 +18,41 @@
       <p>{{event.start}} - {{event.end}}</p>
+      <hr>
+      <h1>Signups</h1>
+      {# This is a link to allow manually navigating to the registration page, but clicking the button only copies the URl to the clipboard. #}
```
`URl` isn't fully capitalized.

Also, how am I supposed to use this link to manually navigate to the page? I'd probably prefer a main link, followed by a smaller `copy link` button

bot190 replied:
Right click, "open on new tab"

Seems like a small use case for a mentor to be at this event page and want to go to the sign up. But yeah the UX might be better if there was a smaller button to copy the link.

bot190 replied:
I've separated these into a normal link as well as a `Copy 🔗` button that copies it to the clipboard.

msoucy commented on signinapp/templates/event_stats.html.jinja2:
```diff
@@ -12,6 +18,41 @@
+            <td>{{block[0].end}}</td>
+            <td>
+              <ul>
+                {%- for registration in block[1] -%}
```
`for user, comment in block[1]`?

msoucy commented on signinapp/templates/event_stats.html.jinja2:
```diff
@@ -12,6 +18,41 @@
+          </tr>
+        </thead>
+        <tbody id="blocksbody">
+        {%- for block in blocks -%}
```
`for times, registrations in blocks`?

bot190 committed:
chore: Small improvements to event factory and event DB calls

Make Event.create() take `str | EventType`
Use `db.session.scalars()` instead of `db.session.execute(QUERY).scalars` throughout events.py

bot190 committed:
fix: Simplify open event ordering

bot190 committed:
fix: Improve event registration handling

Clean-up tuple handing of event registrations
Provide a link to the registration page as well as a copy button
Use bootstrap toast to indicate the link was copied instead of alert.
